### PR TITLE
fix: preserve step sequence in the workflow builder

### LINKED ISSUE #33: Workflow steps are not run in sequence deterministically
### Describe the bug

Workflow builder isn't maintaining the sequence of steps as provided. Internally it maintains a map, and therefore the sequence of steps are lost

### Describe the expected behavior

Workflow builder should ensure the sequence of steps are preserved.

### To Reproduce

The below unit test fails sometimes a Go map doesn't guarantee the ordering:
```
func TestWorkFlowBuilder_MaintainsStepSequence(t *testing.T) {
	b1 := NewStepBuilder("z", WithOnExecute(func(ctx context.Context) (*Report, error) { return &Report{}, nil }))
	b2 := NewStepBuilder("a", WithOnExecute(func(ctx context.Context) (*Report, error) { return &Report{}, nil }))
	b3 := NewStepBuilder("y", WithOnExecute(func(ctx context.Context) (*Report, error) { return &Report{}, nil }))
	b4 := NewStepBuilder("c", WithOnExecute(func(ctx context.Context) (*Report, error) { return &Report{}, nil }))
	b5 := NewStepBuilder("x", WithOnExecute(func(ctx context.Context) (*Report, error) { return &Report{}, nil }))

	wb := NewWorkFlowBuilder("wf").(*workflowBuilder)
	wb.Steps(b1, b2, b3, b4, b5)
	wf, err := wb.Build()
	require.NoError(t, err)
	require.NotNil(t, wf)

	steps := wf.(*workflow).steps
	require.Len(t, steps, 5)
	assert.Equal(t, "z", steps[0].Id())
	assert.Equal(t, "a", steps[1].Id())
	assert.Equal(t, "y", steps[2].Id())
	assert.Equal(t, "c", steps[3].Id())
	assert.Equal(t, "x", steps[4].Id())
}
```

Example output:
```
=== RUN   TestWorkFlowBuilder_MaintainsStepSequence
    workflow_builder_test.go:122: 
        	Error Trace:	/Users/leninmehedy/code/src/github.com/automa-saga/automa/workflow_builder_test.go:122
        	Error:      	Not equal: 
        	            	expected: "z"
        	            	actual  : "a"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-z
        	            	+a
        	Test:       	TestWorkFlowBuilder_MaintainsStepSequence
    workflow_builder_test.go:123: 
        	Error Trace:	/Users/leninmehedy/code/src/github.com/automa-saga/automa/workflow_builder_test.go:123
        	Error:      	Not equal: 
        	            	expected: "a"
        	            	actual  : "y"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-a
        	            	+y
        	Test:       	TestWorkFlowBuilder_MaintainsStepSequence
    workflow_builder_test.go:124: 
        	Error Trace:	/Users/leninmehedy/code/src/github.com/automa-saga/automa/workflow_builder_test.go:124
        	Error:      	Not equal: 
        	            	expected: "y"
        	            	actual  : "c"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-y
        	            	+c
        	Test:       	TestWorkFlowBuilder_MaintainsStepSequence
    workflow_builder_test.go:125: 
        	Error Trace:	/Users/leninmehedy/code/src/github.com/automa-saga/automa/workflow_builder_test.go:125
        	Error:      	Not equal: 
        	            	expected: "c"
        	            	actual  : "x"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-c
        	            	+x
        	Test:       	TestWorkFlowBuilder_MaintainsStepSequence
    workflow_builder_test.go:126: 
        	Error Trace:	/Users/leninmehedy/code/src/github.com/automa-saga/automa/workflow_builder_test.go:126
        	Error:      	Not equal: 
        	            	expected: "x"
        	            	actual  : "z"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-x
        	            	+z
        	Test:       	TestWorkFlowBuilder_MaintainsStepSequence
--- FAIL: TestWorkFlowBuilder_MaintainsStepSequence (0.00s)

Expected :x
Actual   :z
<Click to see difference>


FAIL

Process finished with the exit code 1
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by leninmehedy**
## Description

This pull request updates the workflow builder to ensure that the order in which steps are added is preserved when building workflows. The changes introduce a dedicated sequence to track step insertion order and update relevant methods to use this sequence, with new tests verifying the behavior.

**Step sequence tracking and workflow construction:**

* Added a `stepSequence` field to the `workflowBuilder` struct to record the order steps are added.
* Updated the `Build()` method to construct workflows using the order from `stepSequence` instead of iterating over the unordered `stepBuilders` map.

**Step addition methods:**

* Modified the `Steps()` and `NamedSteps()` methods to initialize and update `stepSequence` whenever steps are added, ensuring the correct order is maintained. [[1]](diffhunk://#diff-e7c112cda9bd742190f9330772f29be4afcc18e33049f3dc0c47553f22d6d0e0R50-R60) [[2]](diffhunk://#diff-e7c112cda9bd742190f9330772f29be4afcc18e33049f3dc0c47553f22d6d0e0R72-R76) [[3]](diffhunk://#diff-e7c112cda9bd742190f9330772f29be4afcc18e33049f3dc0c47553f22d6d0e0R88)

**Testing improvements:**

* Added a new test, `TestWorkFlowBuilder_MaintainsStepSequence`, to verify that the workflow builder preserves the insertion order of steps.

### Related Issues

* Closes #33 
* Related to #

### Pull request (PR) checklist

* \[ ] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[ ] This PR added no TODOs or commented out code
* \[ ] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[ ] Any `go.mod` and/or `go.sum` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[ ] All changes in this PR are included in the description
* \[ ] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit
  message guidelines' below have been followed

### Testing

* \[ ] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* TBD

The following was not tested:

* TBD

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a
   breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any
   type. NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

**Review by brunodam — APPROVED** [review]
Great catch!
